clang/lib/CAS/CASOptions.cpp

```diff
@@ -73,20 +72,17 @@ void CASOptions::freezeConfig(DiagnosticsEngine &Diags) {
     CurrentConfig.CASPath =
         Cache.CAS->getContext().getHashSchemaIdentifier().str();
   }
-  if (Cache.AC)
-    CurrentConfig.CachePath = "";
 }
 
 static std::shared_ptr<llvm::cas::ActionCache>
-createCache(ObjectStore &CAS, const CASConfiguration &Config,
-            DiagnosticsEngine &Diags) {
-  if (Config.CachePath.empty())
+createCacheWithoutPath(const CASConfiguration &Config,
+                       DiagnosticsEngine &Diags) {
+  if (Config.CASPath.empty())
     return llvm::cas::createInMemoryActionCache();
 
+  assert(Config.CASPath == "auto");
   // Compute the path.
-  std::string Path = Config.CachePath;
-  if (Path == "auto")
-    Path = getDefaultOnDiskActionCachePath();
+  std::string Path = getDefaultOnDiskActionCachePath();
```

Comment:
I will ditch this for something new like `getClangDefaultCachePath()`. No need to expose the creation for ObjectStore and ActionCache here.

Reply:
Are you suggesting introducing a `getClangDefaultCachePath()` function which then calls `getDefaultOnDiskActionCachePath()`, instead of calling the latter directly?

Reply:
I think since we decide to bundle object store and action cache together in clang, we should just avoid create them separately. All we need is pick a default path and initialize them together.

Also what do you think if we just get rid of `auto` and InMemory implementation and use empty string as default location?

Reply:
Yeah, that makes sense. We can also replace `CASOptions::getOrCreateObjectStore` and `CASOptions::getOrCreateActionCache` with one call to return both.

Reply:
> Also what do you think if we just get rid of auto and InMemory implementation and use empty string as default location?

After looking into this, it's not straightforward because of 2 reasons:
* There's code like `Options.MCOptions.CAS = CASOpts.getOrCreateObjectStore(Diags);` which after such change would always create an on-disk CAS even if it wasn't used (instead of just using an in-memory one). I don't like that it would be very easy to introduce code that inadvertently opened an on-disk CAS every time clang is invoked.
* For round-tripping of options it becomes ambiguous what an empty `CASPath` means, is the option missing or is the option set to an empty string? I tried with changing `CASPath` to optional string but the `MarshallingInfo` mechanism does not seem able to handle optional fields.

Because of this I kept the in-memory/"auto" functionality for now, we could revisit later on. But I did make other simplifications, see https://github.com/apple/llvm-project/pull/6312/commits/dfc2901bf4ed19e49b2458e976fd3f323312c2cf